src/osd/SnapMapper.cc

```diff
@@ -685,8 +685,12 @@ int SnapMapper::convert_legacy(
       bufferlist bl(iter->value());
       auto bp = bl.cbegin();
       decode(m, bp);
+      string s;
+      s = iter->key().substr(SnapMapper::LEGACY_MAPPING_PREFIX.length());
+      s = SnapMapper::MAPPING_PREFIX + std::to_string(m.hoid.pool) + "_" + s;
+      dout(20) << __func__ << " convert " << iter->key() << " to " << s << dendl;
       to_set.emplace(
-	SnapMapper::get_prefix(m.hoid.pool, m.snap),
```

Comment:
IIUC, `SnapMapper::get_prefix(m.hoid.pool, m.snap) + shard_prefix + m.hoid.to_str()` should return the same output as `s`

Reply:
> Let's fix the get_prefix/convert_legacy helpers rather than bypass them. 

@athanatos - The helpers are not wrong, they use `shard_prefix` and therefore they aren't static, otherwise we could have used `to_raw_key()` instead. Thus we must use the old key for the conversion.

Reply:
Ah, the real problem is that the value contains an encoded Mapping, which contains an hobject rather than a ghobject, so we can't get the shard except by interpreting the string.  This is indeed simpler.